src/main/java/org/spongepowered/api/text/TextArgument.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package org.spongepowered.api.text;
+
+import com.google.common.base.Function;
+import com.google.common.base.Optional;
+
+/**
+ * Represents a function that will be triggered inside {@link Text.Placeholder}
+ * to calculate the replacement for the {@link Text.Placeholder}.
+ */
+public interface TextArgument {
+
+    /**
+     * Resolve the result of TextArgument. If this instance is the main
+     * TextArgumentin a {@link Text.Placeholder} and this method returns
+     * {@link Optional#absent()} the wrapping {@link Text.Placeholder} will not
+     * be replaced.
+     *
+     * @param input The input function that contains the data
+     * @return The replacement for this element or {@link Optional#absent()}
+     *
+     * @see Function#apply(Object)
+     */
+    public Optional<?> apply(Function<String, ?> input);
```

Comment:
> My point was that Function returns `@Nullable`

I will state this more clearly

> I'm also flattered that you are taking stuff out of my PR, but keep in mind that will make them more similar and make yours far less likely to be merged.

Well, our PRs work exactly the same in regards to parameter processing.
Currently the only difference is the serialization. Mine is serializable, yours not (yet).

My `TextArgument`s are basically `Function<Function<String,?>,?>` and Texts.format deals with applying custom any format you would like to use.
Your `TextElement` are reduced to single input params `Function<?,Text>` which basically forces me to use magic or a new processing to apply any customizable formatting. I know this is a little bit unfair, but you don't have your method implemented yet, so I cannot say anything about it.

Let me try to wrap this in an example:
How would you achieve this result:
````
{RED}Some Text {BLUE}Some Other Text.
````
With this given circumstances:
````
Config: {RED}Some Text {BLUE}${varX}.
VarX: Texts.of(TextColors.GREEN, "Some Other Text")
````

My PR:
````java
Text template = stringToTemplate(config);
// Texs.of(TextColors.RED, "Some Text ", TextColors.BLUE, Text.placeholder("varX"), ".");
Map<String,Object> params = Maps.newMap();
params.put("varX", varX);
Text result = Texts.format(template, params);
// Results in: {RED}Some Text {BLUE}Some Other Text.
````

Your PR:
````java
TextTemplate template = stringToTemplate(config);
Text result = template.build(TextArgs.named("varX", varX));
// Results in: ??? (not implemented yet)
````

Maybe we should wait with the discussion until you have finished your impl. Otherwise this is only a theoretical discussion.